### PR TITLE
feat: add capability to identify membrane proxies

### PR DESCRIPTION
**Pull request by caridy**
### Use Case

As a developer, I can add a mark to read-only and read/write proxies so I can identify those proxy objects without attempting to proxify them again.

### API

```js
const flag = Symbol();
const membrane = new ObservableMembrane({
    tagPropertyKey: flag,
});
```

This new configuration called `tagPropertyKey` is optional, and if provided, it is expected to be a `PropertyKey` (string or symbol).

### Semantics

When `tagPropertyKey` configuration is provided, any read-only and read/write proxy created by the membrane will "contain" an own property with that name. E.g.:

```js
flag in proxy; // yields true
Reflect.has(proxy, flag); // yields true
proxy[flag]; // yields undefined
Object.getOwnPropertyDescriptor(proxy, flag); // { value: undefined, writable: false, configurable: false, enumerable: false }
```

### CODE REVIEW

File: src/reactive-handler.ts [hunk 1]
```diff
@@ -98,13 +98,18 @@ export class ReactiveProxyHandler {
...
+        // since key is never going to be undefined, and magicPropertyKey might be undefined
+        // we can simply compare them as the second part of the condition.
+        return key in originalTarget || key === membrane.magicPropertyKey;
     }
     ownKeys(shadowTarget: ReactiveMembraneShadowTarget): string[] {
         const { originalTarget } = this;
-        return ArrayConcat.call(getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
+        const { magicPropertyKey } = this.membrane;
+        const keys = magicPropertyKey ? [magicPropertyKey] : [];
+        return ArrayConcat.call(keys, getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
```
Comment (jdalton) [inline src/reactive-handler.ts:112]: You can call `Reflect.ownKeys(originalTarget)` to get them without doing two separate calls of `getOwnPropertyNames` and `getOwnPropertySymbols`.

File: src/read-only-handler.ts [hunk 2]
```diff
@@ -70,13 +75,17 @@ export class ReadOnlyHandler {
     }
     getOwnPropertyDescriptor(shadowTarget: ReactiveMembraneShadowTarget, key: PropertyKey): PropertyDescriptor | undefined {
         const { originalTarget, membrane } = this;
-        const { valueObserved } = membrane;
+        const { valueObserved, magicPropertyKey } = membrane;
```
Comment (jdalton) [inline src/read-only-handler.ts:79]: Add a test for the `o[flag]` is `undefined` result and then one for getting the descriptor.

**Comment by jdalton** [comment]
> Object.getOwnPropertyDescriptor(proxy, flag);

What do you think about making it not show up in `Object.getOwnPropertyDescriptor()` or `Reflect.ownKeys()` this so it only shows up in `in` and `Reflect.has` operations?

File: src/reactive-membrane.ts [hunk 3]
```diff
@@ -35,15 +36,20 @@ export interface ObservableMembraneInit {
...
+function createShadowTarget(value: any, magicPropertyKey: PropertyKey | undefined): ReactiveMembraneShadowTarget {
     let shadowTarget: ReactiveMembraneShadowTarget | undefined = undefined;
     if (isArray(value)) {
         shadowTarget = [];
     } else if (isObject(value)) {
         shadowTarget = {};
     }
+    if (!isUndefined(magicPropertyKey)) {
+        // installing the magic property as non-configurable, non-writable and non-enumerable
+        ObjectDefineProperty(shadowTarget, magicPropertyKey, ObjectCreate(null));
```
Comment (dgitin) [inline src/reactive-membrane.ts:51]: Have you considered making the property `writable: true` ? That way it could cover more use cases.

File: src/reactive-handler.ts (hunk 1 above)
Comment (caridy) [reply src/reactive-handler.ts:112]: This lib is written in ways to support IE11, so we stick to the basis.

File: src/read-only-handler.ts (hunk 2 above)
Comment (caridy) [reply src/read-only-handler.ts:79]: yeah... tests are coming...

**Comment by caridy** [comment]
@jdalton 

> What do you think about making it not show up in Object.getOwnPropertyDescriptor() or Reflect.ownKeys() this so it only shows up in in and Reflect.has operations?

that breaks the object invariants if they decide to freeze the proxy. That's why we have to provide a consistent result for `has`, `getOwnPropertyDescriptor` and `ownKeys`.

@dgitin 

> Would making the property writable and available via set be too much ? I'm thinking this could be used for other scenarios when a simple true/false check may not be enough.

do you have an actual use-case here? I will like to stick the MVP here.

File: src/reactive-handler.ts (hunk 1 above)
Comment (jdalton) [reply src/reactive-handler.ts:112]: Instead of concat you can use something like `ArrayPush` since you don't need to create a new array:
```js
ArrayPush.apply(keys, getOwnPropertyNames(originalTarget));
ArrayPush.apply(keys, getOwnPropertySymbols(originalTarget));
```

Comment (caridy) [reply src/reactive-handler.ts:112]: will do that in another PR...

Comment (jdalton) [reply src/reactive-handler.ts:112]: Cool, things LGTM

Comment (caridy) [reply src/reactive-handler.ts]: will do

File: src/read-only-handler.ts (hunk 2 above)
Comment (caridy) [reply src/read-only-handler.ts:74]: tests are in...

File: src/reactive-dev-formatter.ts [hunk 4]
```diff
@@ -31,9 +29,8 @@ function extract(objectOrArray: any): any {
     }
 
     const obj = ObjectCreate(getPrototypeOf(objectOrArray));
-    const names = getOwnPropertyNames(objectOrArray);
-    return ArrayConcat.call(names, getOwnPropertySymbols(objectOrArray))
-        .reduce((seed: any, key: PropertyKey) => {
+    const names = getOwnKeysAndMagicProperty(objectOrArray, undefined);
```
Comment (caridy) [inline src/reactive-dev-formatter.ts:32]: just reusing a new shared lib function

File: src/reactive-handler.ts [hunk 5]
```diff
@@ -37,21 +35,21 @@ function unwrapDescriptor(descriptor: PropertyDescriptor): PropertyDescriptor {
 }
 
 function lockShadowTarget(membrane: ReactiveMembrane, shadowTarget: ReactiveMembraneShadowTarget, originalTarget: any): void {
-    const targetKeys = ArrayConcat.call(getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
+    const targetKeys = getOwnKeysAndMagicProperty(originalTarget, membrane.magicPropertyKey);
     targetKeys.forEach((key: PropertyKey) => {
-        let descriptor = getOwnPropertyDescriptor(originalTarget, key) as PropertyDescriptor;
```
Comment (caridy) [inline src/reactive-handler.ts:42]: removing the casting... in favor of an if condition, this also help with the case were one key is not in the original target because it is the magic key

File: src/reactive-handler.ts [hunk 6]
```diff
@@ -98,13 +96,15 @@ export class ReactiveProxyHandler {
...
-        return key in originalTarget;
+        // since key is never going to be undefined, and magicPropertyKey might be undefined
+        // we can simply compare them as the second part of the condition.
+        return key in originalTarget || key === membrane.magicPropertyKey;
     }
-    ownKeys(shadowTarget: ReactiveMembraneShadowTarget): string[] {
-        const { originalTarget } = this;
-        return ArrayConcat.call(getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
+    ownKeys(shadowTarget: ReactiveMembraneShadowTarget): PropertyKey[] {
+        return getOwnKeysAndMagicProperty(this.originalTarget, this.membrane.magicPropertyKey);
```
Comment (caridy) [inline src/reactive-handler.ts:107]: reusing shared lib func

File: src/reactive-handler.ts [hunk 7]
```diff
@@ -178,11 +182,16 @@ export class ReactiveProxyHandler {
         // if the descriptor has a value, as opposed to getter/setter
         // So we can just check if writable is present and then see if
         // value is present. This eliminates getter and setter descriptors
-        if (hasOwnProperty.call(descriptor, 'writable') && !hasOwnProperty.call(descriptor, 'value')) {
-            const originalDescriptor = getOwnPropertyDescriptor(originalTarget, key) as PropertyDescriptor;
+        const originalDescriptor = getOwnPropertyDescriptor(originalTarget, key);
```
Comment (caridy) [inline src/reactive-handler.ts:185]: this change is needed to accommodate for the new fact that sometimes the key is not installed on the original target, but on the shadow for the magic key

File: src/reactive-membrane.ts [hunk 8]
```diff
@@ -35,15 +35,20 @@ export interface ObservableMembraneInit {
...
+function createShadowTarget(value: any, magicPropertyKey: PropertyKey | undefined): ReactiveMembraneShadowTarget {
     let shadowTarget: ReactiveMembraneShadowTarget | undefined = undefined;
     if (isArray(value)) {
         shadowTarget = [];
     } else if (isObject(value)) {
         shadowTarget = {};
     }
+    if (!isUndefined(magicPropertyKey)) {
+        // installing the magic property as configurable, non-writable and non-enumerable
+        ObjectDefineProperty(shadowTarget, magicPropertyKey, { configurable: true });
```
Comment (caridy) [inline src/reactive-membrane.ts:50]: configurable, non-writable is the optimal definition since it can still change if needed, and the object invariants will not apply

File: src/shared.ts [hunk 9]
```diff
@@ -4,7 +4,6 @@ const {
     getPrototypeOf,
     create: ObjectCreate,
     defineProperty: ObjectDefineProperty,
-    defineProperties: ObjectDefineProperties,
```
Comment (caridy) [inline src/shared.ts:7]: removing tons of stuff that were not used anywhere

File: src/shared.ts [hunk 10]
```diff
@@ -66,3 +58,11 @@ export function registerProxy(proxy: object, value: any) {
 }
 
 export const unwrap = (replicaOrAny: any): any => proxyToValueMap.get(replicaOrAny) || replicaOrAny;
+
+export function getOwnKeysAndMagicProperty(target: object, magicPropertyKey: PropertyKey | undefined): PropertyKey[] {
+    const keys = isUndefined(magicPropertyKey) || hasOwnProperty.call(target, magicPropertyKey) ? [] : [magicPropertyKey];
```
Comment (caridy) [inline src/shared.ts:63]: this method follows @jdalton suggestion to avoid the creation of an extra array when doing concat. it also adds the magic key when present.

File: src/reactive-dev-formatter.ts [hunk 11]
```diff
@@ -31,9 +29,8 @@ function extract(objectOrArray: any): any {
...
+    return names.reduce((seed: any, key: PropertyKey) => {
             const item = objectOrArray[key];
```
Comment (jdalton) [inline src/reactive-dev-formatter.ts:34]: Could add `ArrayReduce`

**Comment by jdalton** [comment]
I saw a code comment about freezing objects. Is there complications there?

**Comment by caridy** [comment]
@jdalton 

> I saw a code comment about freezing objects. Is there complications there?

Yes, the problem is that when freezing the proxy, the engine will read all own keys, then loop them to get each individual descriptor (this list will contain the magic key), then it calls `defineProperty`, which we normally install on the actual target, this will leak the magic key as a descriptor in the real target. I added a condition to avoid leaking that particular magic key.

**Comment by pmdartus** [comment]
Bikeshedding: While I like the magic aspect of the `magicPropertyKey`, what do you think about naming it something like `membaneStamp` or `membraneKey`?

**Comment by jdalton** [comment]
@caridy 
> > I saw a code comment about freezing objects. Is there complications there?
> 
> Yes, the problem is that when freezing the proxy, the engine will read all own keys, then loop them to get each individual descriptor (this list will contain the magic key), then it calls `defineProperty`, which we normally install on the actual target, this will leak the magic key as a descriptor in the real target. I added a condition to avoid leaking that particular magic key.

Will projecting the property on a frozen object cause problems if the property is not installed on the object?

@pmdartus 
> Bikeshedding: While I like the magic aspect of the `magicPropertyKey`, what do you think about naming it something like `membraneStamp` or `membraneKey`?

I'll throw in `membraneBrand` or `membraneTag`.

**Comment by caridy** [comment]
I like `membraneTag`

@jdalton 
> Will projecting the property on a frozen object cause problems (errors) if the property is not installed on the object?

The property is on the shadowTarget, if you freeze the proxy, it freezes the target and the shadowTarget, the target will not get the magic key, while the shadowTarget has it.

File: src/reactive-handler.ts [hunk 12]
```diff
@@ -178,11 +182,16 @@ export class ReactiveProxyHandler {
...
+        if (!isUndefined(originalDescriptor) && hasOwnProperty.call(descriptor, 'writable') && !hasOwnProperty.call(descriptor, 'value')) {
```
Comment (leobalter) [inline src/reactive-handler.ts:186]:
```suggestion
        if (originalDescriptor && hasOwnProperty.call(descriptor, 'writable') && !hasOwnProperty.call(descriptor, 'value')) {
```

File: src/reactive-handler.ts [hunk 13]
```diff
@@ -178,11 +182,16 @@ export class ReactiveProxyHandler {
...
             descriptor.value = originalDescriptor.value;
         }
-        ObjectDefineProperty(originalTarget, key, unwrapDescriptor(descriptor));
+        // we must avoid installing the magic descriptor into the original target,
+        // because it will leak internal implementation details. This is tricky because
+        // Object.freeze() will attempt to do so.
+        if (magicPropertyKey !== key || !isUndefined(originalDescriptor)) {
```
Comment (leobalter) [inline src/reactive-handler.ts:192]: same here for !isUndefined

File: src/reactive-membrane.ts [hunk 14]
```diff
@@ -35,15 +35,20 @@ export interface ObservableMembraneInit {
...
 
-function createShadowTarget(value: any): ReactiveMembraneShadowTarget {
+function createShadowTarget(value: any, magicPropertyKey: PropertyKey | undefined): ReactiveMembraneShadowTarget {
     let shadowTarget: ReactiveMembraneShadowTarget | undefined = undefined;
     if (isArray(value)) {
         shadowTarget = [];
     } else if (isObject(value)) {
         shadowTarget = {};
     }
+    if (!isUndefined(magicPropertyKey)) {
```
Comment (leobalter) [inline src/reactive-membrane.ts:48]: these `!isUndefined` checks are weird. For the property key I'd suggest creating a method for: isPropertyKey following the [semantics of ES](https://tc39.es/ecma262/#sec-ispropertykey):

1. If Type(argument) is String, return true.
1. If Type(argument) is Symbol, return true.
1. Return false.

File: src/read-only-handler.ts [hunk 15]
```diff
@@ -70,13 +70,17 @@ export class ReadOnlyHandler {
...
 
         // keys looked up via hasOwnProperty need to be reactive
         valueObserved(originalTarget, key);
 
         let desc = getOwnPropertyDescriptor(originalTarget, key);
         if (isUndefined(desc)) {
+            if (key === magicPropertyKey) {
+                // key is the magic key, the shadow target must have a descriptor for it.
```
Comment (leobalter) [inline src/read-only-handler.ts:81]:
```suggestion
                // the shadow target must have a descriptor this key
```

File: src/shared.ts (hunk 10 above)
Comment (leobalter) [inline src/shared.ts:63]: `isUndefined` returns a boolean. I'm not sure if `const keys = isUndefined(magicPropertyKey)` can be problematic.

```suggestion
    if (isUndefined(magicPropertyKey)) {
        return undefined;
    }
    const keys = hasOwnProperty.call(target, magicPropertyKey) ? [] : [magicPropertyKey];
```

Particularly, I don't love ternaries so I'd also avoid them:

```suggestion
    if (isUndefined(magicPropertyKey)) {
        return undefined;
    }
    const keys = [];
    if (!hasOwnProperty.call(target, magicPropertyKey)) {
      ArrayPush.apply(keys, magicPropertyKey);
    }
```

Problem here is the `!`, but it could be replaced with a trailing `=== false`

File: src/shared.ts [hunk 16]
```diff
@@ -66,3 +58,11 @@ export function registerProxy(proxy: object, value: any) {
 }
 
 export const unwrap = (replicaOrAny: any): any => proxyToValueMap.get(replicaOrAny) || replicaOrAny;
+
+export function getOwnKeysAndMagicProperty(target: object, magicPropertyKey: PropertyKey | undefined): PropertyKey[] {
```
Comment (leobalter) [inline src/shared.ts:62]: This is missing tests in the shared.spec.ts file.

File: src/reactive-handler.ts [hunk 17]
```diff
@@ -37,21 +35,21 @@ function unwrapDescriptor(descriptor: PropertyDescriptor): PropertyDescriptor {
...
-
-        // We do not need to wrap the descriptor if configurable
-        // Because we can deal with wrapping it when user goes through
-        // Get own property descriptor. There is also a chance that this descriptor
-        // could change sometime in the future, so we can defer wrapping
-        // until we need to
-        if (!descriptor.configurable) {
-            descriptor = wrapDescriptor(membrane, descriptor, wrapValue);
+        let descriptor = getOwnPropertyDescriptor(originalTarget, key);
+        if (!isUndefined(descriptor)) {
```
[the remark this answers is not in the document]
Comment (caridy) [reply src/reactive-handler.ts:41]: that's the thing, `getOwnKeysAndMagicProperty` will add a the magical key to the list... which might or might not be in the originalTarget.

File: src/reactive-handler.ts (hunk 12 above)
Comment (caridy) [reply src/reactive-handler.ts:186]: all these libraries are using `isUndefined` helper for two reasons:

* it is supposed to be faster because it doesn't have to cast the value.
* It is easier to read and reasoning about what the condition is about.

File: src/reactive-membrane.ts (hunk 14 above)
Comment (caridy) [reply src/reactive-membrane.ts:48]: the `isUndefined` is weird, yes... ideally we can have `isDefined` but I don't think TS allows to do a good declaration of that, @pmdartus might know how... but in general, we use isUndefined everywhere... see my previous message.

In this case, it is because the magical property key, which is well defined in the type system, could be undefined or a valid property key.

File: src/shared.ts (hunk 10 above)
Comment (caridy) [reply src/shared.ts:63]: there are probably some gaps here... but based on the signature of this, you must return an array, not undefined... and that array should alwyas contain all the own properties of the target, and one extra property, the magic property, if the target doesnt' have it.

Comment (leobalter) [reply src/shared.ts:63]: I really got confused with the ternary. Now I see it gets the `[]` if magicPropertyKey is undefined.

File: src/reactive-membrane.ts (hunk 8 above)
Comment (pmdartus) [reply src/reactive-membrane.ts:50]: When would we reuse the same shadow target and reset the property?

File: test/reactive-handler.spec.ts [hunk 18]
```diff
@@ -791,4 +791,55 @@ describe('ReactiveHandler', () => {
...
+            });
+
+            const wet = target.getProxy(o);
+            Object.freeze(wet);
+            expect(Object.getOwnPropertyNames(wet).length).toBe(1);
+            expect('foo' in wet).toBe(true);
+            expect('foo' in o).toBe(false);
+            expect(wet.foo).toBe(undefined);
+        });
+    });
```
Comment (ravijayaramappa) [inline test/reactive-handler.spec.ts:844]: Add test cases for 
1. Defining a new property using Object.defineProperty on the wet object
2. Delete the magic property key on the wet object
3. Have magic property defined on dry object, define same property on wet object, delete the property on wet object. Recheck if property exists and the value.

File: src/reactive-membrane.ts (hunk 8 above)
Comment (caridy) [reply src/reactive-membrane.ts:50]: we will never reuse a shadow target, is that what you're asking?

File: src/reactive-dev-formatter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (caridy) [reply src/reactive-dev-formatter.ts:32]: that will be a type error because the argument is expect. we can make it optional though.

File: src/shared.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (caridy) [reply src/shared.ts:63]: discussed offline... we are good here... the target might or might not have such key.

File: test/reactive-handler.spec.ts (hunk 18 above)
Comment (pmdartus) [reply test/reactive-handler.spec.ts:844]: On top of this list, I would also add test to check the returned property descriptor of the magic propety on the wet object: `{ value: undefined, enumerable: false, configurable: false, writable: false }`

File: src/reactive-membrane.ts (hunk 8 above)
Comment (pmdartus) [reply src/reactive-membrane.ts:50]: Yes. I don't understand why the `magicPropertyKey` is set to configurable.

File: src/base-handler.ts [hunk 19]
```diff
@@ -9,17 +8,14 @@ import {
...
-    isArray,
     freeze,
+    ArrayPush,
+    ObjectCreate,
 } from './shared';
 import { ReactiveMembrane } from './reactive-membrane';
 
 export type ReactiveMembraneShadowTarget = object;
 
-export function createShadowTarget(value: any): ReactiveMembraneShadowTarget {
```
Comment (caridy) [inline src/base-handler.ts:19]: this was only used by membrane, moving it there.

File: src/base-handler.ts [hunk 20]
```diff
@@ -66,10 +62,17 @@ export abstract class BaseProxyHandler {
     }
     lockShadowTarget(shadowTarget: ReactiveMembraneShadowTarget): void {
         const { originalTarget } = this;
-        const targetKeys = ArrayConcat.call(getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
```
Comment (caridy) [inline src/base-handler.ts:69]: removing the concat in favor of push, not material to this PR

File: src/reactive-membrane.ts [hunk 21]
```diff
@@ -62,20 +62,26 @@ const defaultValueMutated: ReactiveMembraneMutationCallback = (obj: any, key: Pr
 };
 const defaultValueDistortion: ReactiveMembraneDistortionCallback = (value: any) => value;
 
+function createShadowTarget(value: any): any {
```
Comment (caridy) [inline src/reactive-membrane.ts:65]: moved from base since it is only used here

File: test/reactive-handler.spec.ts [hunk 22]
```diff
@@ -873,12 +873,62 @@ describe('ReactiveHandler', () => {
             it('should notify on pop()', () => {
                 const value = ['foo', 'bar'];
                 const proxy = membrane.getProxy(value);
-                debugger;
```
Comment (caridy) [inline test/reactive-handler.spec.ts:876]: someone left this in here :p

File: test/reactive-handler.spec.ts (hunk 18 above)
Comment (caridy) [reply test/reactive-handler.spec.ts:942]: @pmdartus this is the test that you asked for, in fact surface a problem, thanks:
https://github.com/salesforce/observable-membrane/pull/48/commits/b2aec8c5695fcbee8e3b1484762b78ad5c9bb215

@ravijayaramappa I don't understand your test... can you elaborate?

File: test/reactive-handler.spec.ts [hunk 23]
```diff
@@ -873,12 +873,71 @@ describe('ReactiveHandler', () => {
...
                 expect(proxy.pop()).toBe('bar');
                 expect(changeSpy).toHaveBeenCalledTimes(2);
                 expect(changeSpy).toHaveBeenCalledWith(value, 'length');
                 expect(changeSpy).toHaveBeenCalledWith(value, '1');
             });
         });
     });
+    describe('with magic key property', () => {
```
Comment (jdalton) [inline test/reactive-handler.spec.ts:883]: magic key -> tag key property

**Comment by jdalton** [comment]
LGTM. The freezing issue is resolved it looks like too?

**Comment by caridy** [comment]
> LGTM. The freezing issue is resolved it looks like too?

Yes!

File: test/reactive-handler.spec.ts (hunk 18 above)
Comment (ravijayaramappa) [reply test/reactive-handler.spec.ts:942]: Assuming that the magic property is discoverable, i was suggesting to add test cases for the corner cases.
```js
const o = {};
const target = new ReactiveMembrane({
    tagPropertyKey: '$$MagicKey$$',
});

const wet = target.getProxy(o);
Object.defineProperty(wet, '$$MagicKey$$', {
    value: 'bar',
    configurable: true
});

#1
expect('$$MagicKey$$' in wet).toBe(true);
expect(object.keys(wet)).toEqual(['$$MagicKey$$']);

#2
delete wet["$$MagicKey$$"]
expect('$$MagicKey$$' in wet).toBe(false); // what should happen here?
expect(object.keys(wet)).toEqual([]); 
```

Comment (ravijayaramappa) [reply test/reactive-handler.spec.ts:942]:
```js
#3
const dry = {};
const target = new ReactiveMembrane({
    tagPropertyKey: '$$MagicKey$$',
});

const wet = target.getProxy(dry);
Object.defineProperty(dry, '$$MagicKey$$', {
    value: 'bar',
    configurable: true
});

Object.defineProperty(wet, '$$MagicKey$$', {
    value: 'baz',
    configurable: true
});

expect('$$MagicKey$$' in wet).toBe(true);
expect(object.keys(wet)).toEqual(['$$MagicKey$$']);
expect(wet['$$MagicKey$$]).toBe('baz');
expect(dry['$$MagicKey$$]).toBe('baz'); // value set on wet propagated to dry

delete wet["$$MagicKey$$"]
expect('$$MagicKey$$' in wet).toBe(false);
expect(object.keys(wet)).toEqual([]);
expect(dry['$$MagicKey$$]).toBeUndefined();

expect('$$MagicKey$$' in dry).toBe(false);
expect(object.keys(dry)).toEqual([]);
expect(dry['$$MagicKey$$]).toBeUndefined();
```

#4 would be the same test setup with the descriptor marked as `configurable: false`

File: src/base-handler.ts [hunk 24]
```diff
@@ -121,16 +129,21 @@ export abstract class BaseProxyHandler {
...
+        let desc = getOwnPropertyDescriptor(originalTarget, key);
         if (isUndefined(desc)) {
-            return undefined;
+            if (key !== tagPropertyKey) {
+                return undefined;
+            }
+            // if the key is the membrane tag key, and is not in the original target,
+            // we produce a synthetic descriptor and install it on the shadow target
+            desc = { value: undefined, writable: false, configurable: false, enumerable: false };
+            ObjectDefineProperty(shadowTarget, tagPropertyKey, desc);
```
Comment (ravijayaramappa) [inline src/base-handler.ts:145]: It is safe to return the `desc` at this point. This one is pretty harmless and does not need to be wrapped.
